File: engine/Utils/dataTypes.py
from enum import Enum
from os.path import join as os_join
from datetime import date

class Counter(dict):
    def union(self, counter):
        for key,value in counter.items():
            self.add(key, value)

    def add(self, name, number=1):
        if(name in self):
            self[name] += number
        else:
            self[name] = number
        
    def remove(self, name, number=1):
        if(name in self):
            self[name] -= number
            if(self[name] == 0):
                del self[name]
    
    def decrease(self, name, number=1):
        if(self.get(name, 0)>0): self[name] -= number
    
    def decreaseAll(self, number=1):
        for key, value in self.items():
            if(value>0): self[key] = value - number

    def isEmpty(self):
        return True if len(self.keys) == 0 else False
    
    def getAllEqualOrAbove(self, number):
        return [ key for key,value in self.items() if (value>=number)]
    
    def getKeyWithHigherValue(self):
        return max(self, key=self.get, default=None)
    
    def to_s(self):
        return [ f"{key}({value})" for key,value in self.items()]

class FrequencyCounter(Counter):
    def __init__(self):
        self.mostFrequentKey = None
    
    def setMostFrequentKey(self):
        self.mostFrequentKey = self.getKeyWithHigherValue()
    
    def updateMostFrequentKey(self, lastUpdatedKey):
        if(self.mostFrequentKey != lastUpdatedKey and self.get(lastUpdatedKey, 0)>self.get(self.mostFrequentKey, 0)):
            self.mostFrequentKey = lastUpdatedKey

    def add(self, name, number=1):
        super().add(name, number)
        self.updateMostFrequentKey(name)
        
    def remove(self, name, number=1):
        super().remove(name, number)
        if(name == self.mostFrequentKey): self.setMostFrequentKey()

class ArticleTagCounter(Counter):
    def merge(self, articleCounter):
        self.union(articleCounter)
    
    def verify(self, minTagsForAcceptance, maxTagForAcceptance):
        if(len(self.keys())>=minTagsForAcceptance and self.get("img", 0) == 1 and (self.get("span", 0)+self.get("text", 0)) < 6 and self.get("div", 0) < 4 and self.lazyCount(maxTagForAcceptance)):
            return True
        else: return False
    
    def clear(self):
        if(len(self)>0):
            super().clear()
    
    def lazyCount(self, max):
        sum = 0
        for value in self.values():
            sum += value
            if(sum>max): return False
        return True

class InterpreterEnvironment:
    def __init__(self, minTagsForAcceptance=3, minNumOccurences=5, maxTagsForAcceptance=10):
        self.tagCounter = ArticleTagCounter()
        self.articleClasses = Counter()
        self.minNumOccurences = minNumOccurences
        self.minTagsForAcceptance = minTagsForAcceptance
        self.maxTagsForAcceptance = maxTagsForAcceptance
    
    def resetTagCounter(self):
        if(len(self.tagCounter.keys())>0):
            self.tagCounter = ArticleTagCounter()
    
    def checkArticleClasses(self):
        newArticleClasses = Counter()

        self.articleClasses.pop(None, None)

        for articleClass, count in self.articleClasses.items():
            if(count>=self.minNumOccurences): newArticleClasses.add(articleClass, count)
        
        self.articleClasses = newArticleClasses
    
    def createNewsWebsiteInfo(self, url, name, domain, faviconName, maxFailedAttemps=1):
        return NewsWebsiteInfo(url, name, domain, faviconName, set(self.articleClasses.keys()), maxFailedAttemps)

class WebsiteInfo:
    def __init__(self, url, name, url_id, website_id):
        self.website_id = website_id
        self.url = url
        self.name = name
        self.url_id = url_id
        
class NewsWebsiteInfo(WebsiteInfo):
    def __init__(self, url, name, domain, faviconName, articleClasses = set(), maxFailedAttemps=1):
        super().__init__(url, name, domain, faviconName)
        self.maxFailedAttemps = maxFailedAttemps
        self.tags = Counter()
        self.invalidTags = set()
        self.articleClasses = articleClasses
        self.articleElements = {}
        self.lastTagsUpdate = date.today()
    
    def isEmpty(self):
        return len(self.tags) == 0

class NewsWebsitesInfo:
    def __init__(self):
        self.nonNewsWebsites = set(["https://www.google.com", "https://www.bing.com", "https://duckduckgo.com/", "https://www.yahoo.com"])
        self.newsWebsitesInfo = {}
        self.newsWebsitesScores = Counter()

class LazySplitter:
    def __init__(self, word, delimiter):
        self.word = word + "" if word[-1] == delimiter else delimiter
        self.delimiter = delimiter
        self.index = 0

    def next(self):
        end = self.word.find(self.delimiter, self.index)
        if(end != -1): 
            word = self.word[self.index:end]
            self.index = end+1
            return word
        else: return None

class URLLazySplitter:
    def __init__(self, word):
        self.delimiter = "/"
        self.word = word + "" if word[-1] == self.delimiter else self.delimiter
        self.index = self._getFirstIndex()

    def _getFirstIndex(self):
        index  = self.word.find("//")
        if(index != -1):
            return index + 2
        else: return 0
    
    def reset(self, word):
        self.word = word + "" if word[-1] == self.delimiter else self.delimiter
        self.index = self._getFirstIndex()

    def next(self):
        end = self.word.find(self.delimiter, self.index)
        if(end != -1): 
            word = self.word[self.index:end]
            self.index = end+1
            return word
        else: return None

class ArticleItems(Enum):
    IMG = 0
    HEADER = 1
    TITLE = 2
    DESCRIPTION = 3
    CODE = 4

def createArticleElementsTable(img=None, header=None, title=None, description=None):
    return {ArticleItems.IMG: img, ArticleItems.HEADER: header, ArticleItems.TITLE: title, ArticleItems.DESCRIPTION: description}

class LabeledObject:
    def __init__(self, label, obj):
        self.label = label
        self.object = obj

class DatabaseInfo():
    def __init__(self, path, databaseFile, browser):
        self.path = path
        self.databaseFile = databaseFile
        self.browser = browser
    
    def get_path(self):
        return os_join(self.path, self.databaseFile)

def createScheduleObj(title=None, date=None, startTime=None, endTime=None):
    return {
        ScheduleOperators.TITLE: title,
        ScheduleOperators.DATE: date,
        ScheduleOperators.TIME: [startTime, endTime]
    }

class ScheduleOperators(Enum):
    NONE = 0
    TITLE = 1
    TIME = 2
    DATE = 3
    INTERVAL = 4
    ACTION = 5

class ScheduleActions(Enum):
    ADD = 0
    DEL = 1
    MOD = 2
    NONE = 3

File: gui/linkers/modules/utils/dataTypes.js
import { WEBSITE_TYPES } from "../consts/websiteTypes.js"
import { get } from "./utils.js"
import { treat_all_url } from "./websiteUtils.js"

class Counter{
    constructor(values={}){
        this.values = values
        this.size = 0
    }

    has = (key) => key in this.values

    clear = () => {this.values = {}}

    keys = () => Object.keys(this.values)

    pop = (key) => {
        let value = this.values[key]
        if(value !== undefined){
            this.size -= value
            delete this.values[key]
        }
    }

    union = (counter) => {
        var otherValues = Object.entries(counter.values)
        let lenOtherValues = otherValues.length

        for(var i=0; i<lenOtherValues; i++){
            this.add(...otherValues[i])
        }
    }

    add = (name, number=1) => {
        if(name in this.values){
            this.values[name] += number
        }else{
            this.values[name] = number
        }

        this.size += number
    }
        
    remove = (name, number=1) => {
        if(name in this.values){
            this.values[name] -= number
            if(this.values[name] === 0)
                delete this.values[name]
        }

        this.size -= number
    }
    
    decrease = (name, number=1) => {
        if(get(this.values, name, 0)>0) this[name] -= number

        this.size -= number
    }

    decreaseAll = (number=1) => {
        let keys = this.values.key()
        let lenOtherValues = keys.length

        for(var i=0; i<lenOtherValues; i++){
            if(this.values[keys[i]]>0){
                this.values[keys[i]] -= number
                this.size -= number
            }
        }
    }

    isEmpty = () => {
        return this.values.length === 0
    }
    
    getAllEqualOrAbove = (number) => {
        var keys = []
        let entries = Object.entries(this.values)
        let lenEntries = entries.length

        for(var i=0; i<lenEntries; i++){
            let [key, value] = entries[i]
            if(value>number) keys.push(key)
        }

        return keys
    }
    
    getKeyWithHigherValue = () => {
        let entries = Object.entries(this.values)
        let len = entries.length
        if(len>1){
            let keyWithHigherValue = null
            let higherValue = null
            let times = 0
        
            for(var i=0; i<len; i++){
                let [key, value] = entries[i]

                if(value>higherValue){
                    keyWithHigherValue = key
                    higherValue = value
                    times = 1
                }else if(value === higherValue) times += 1
            }

            return times == 1? keyWithHigherValue: null
        }else return len === 1? entries[0][0]: null
    }
    
    toString = () => {
        let entries = Object.entries(this.values)
        let len = entries.length
        let arr = []

        for(var i=0; i<len; i++){
            let [key, value] = entries[i]
            arr.push(`${key}(${value})`)
        }

        return arr
    }

    getSortedArray = () => Object.entries(this.values).sort(([,a], [,b]) => b-a)
}

class FrequencyCounter extends Counter{
    constructor(){
        this.mostFrequentKey = null
    }
    
    setMostFrequentKey = () => {
        this.mostFrequentKey = this.getKeyWithHigherValue()
    }
    
    updateMostFrequentKey = (lastUpdatedKey) => {
        if(this.mostFrequentKey != lastUpdatedKey && get(this.values, lastUpdatedKey, 0)>get(this.values, this.mostFrequentKey, 0))
            this.mostFrequentKey = lastUpdatedKey
    }

    add = (name, number=1) => {
        super.add(name, number)
        this.updateMostFrequentKey(name)
    }
        
    remove = (name, number=1) => {
        super.remove(name, number)
        if(name === this.mostFrequentKey) this.setMostFrequentKey()
    }
}

class ArticleTagCounter extends Counter{
    merge = (articleCounter) => {
        this.union(articleCounter)
    }
    
    verify = (decisionTree, acceptancePercentage=0.5) => decisionTree.evaluate(this.values) >= acceptancePercentage

    clear = () => {
        if(Object.keys(this.values).length>0){
            this.values = {}
            this.length = 0
        }
    }
    
    lazyCount = (max, min=0, minNumTags=2) => {
        delete this.values.undefined
        let keys = Object.keys(this.values)
        let len = keys.length
        
        if(len<minNumTags) return false
        let sum = 0
        
        for(var i=0; i<len; i++){
            sum += this.values[keys[i]]
            if(sum>max) return false
        }

        return sum>=min
    }
}

class InterpreterEnvironment{
    constructor(minTagsForAcceptance=3, minNumOccurences=5, maxTagsForAcceptance=10, decisionTree=getNewsDecisionTree(), acceptancePercentage=0.8){
        this.tagCounter = new ArticleTagCounter()
        this.articleClasses = new Counter()
        this.minNumOccurences = minNumOccurences
        this.minTagsForAcceptance = minTagsForAcceptance
        this.maxTagsForAcceptance = maxTagsForAcceptance
        this.decisionTree = decisionTree
        this.acceptancePercentage = acceptancePercentage
    }
    
    resetTagCounter = () => {
        if(Object.keys(this.tagCounter.values).length>0){
            this.tagCounter = new ArticleTagCounter()
        }
    }
    
    checkArticleClasses = () => {
        delete this.articleClasses.values.null

        let entries = Object.entries(this.articleClasses.values)
        let length = entries.length

        for(var i=0; i<length; i++){
            var [articleClass, count] = entries[i]
            if(count<this.minNumOccurences){
                delete this.articleClasses.values[articleClass]  
            } 
        }
    }
    
    createNewsWebsiteInfo = (url, maxFailedAttemps=1) => {
        return new NewsWebsiteInfo(undefined, url, undefined, new Set(Object.keys(this.articleClasses.values)), undefined, maxFailedAttemps)
    }

    getArticleClasses = () => new Set(this.articleClasses.keys())
}

class WebsiteInfo{
    constructor(id, url, domain, name){
        this.id = id
        this.url = url
        this.name = name
        this.domain = domain
    }
}
        
class NewsWebsiteInfo extends WebsiteInfo{
    constructor(id="", url, name="", type = WEBSITE_TYPES.GENERAL, score = 0, articlesInfo = {}, lastTagsUpdate = moment(), maxFailedAttemps=1){
        super(id, url, "", name)
        this.maxFailedAttemps = maxFailedAttemps
        this.categories = {}
        this.invalidCategories = new Set()
        this.articlesInfo = articlesInfo
        this.lastTagsUpdate = lastTagsUpdate
        this.score = score
        this.type = type
    }
    
    isEmpty = () => {
        return Object.keys(this.categories.values).length == 0
    }

    addCategory = (url, obj) => {
        this.categories[url] = obj
    }

    static fromObject = (obj, url=undefined) => {
        return new NewsWebsiteInfo(obj.id, url?url:treat_all_url(obj.url), obj.name, obj.type, obj.score, undefined, moment.unix(obj.last_update))
    }
}

class NewsWebsitesInfo{
    constructor(){
        this.nonNewsWebsites = new Set(["https://www.google.com", "https://www.bing.com", "https://duckduckgo.com/", "https://www.yahoo.com"])
        this.newsWebsitesInfo = {}
        this.newsWebsitesScores = new Counter()
    }
}

class LazySplitter{
    index = 0

    constructor(word, delimiter){
        this.word = `${word}${word[-1]==delimiter?"":delimiter}`
        this.delimiter = delimiter
    }

    next = () => {
        let end = this.word.find(this.delimiter, this.index)
        if(end != -1){
            let word = this.word.slice(this.index, end)
            this.index = end+1
            return word
        }else return null
    }
}

class URLLazySplitter extends LazySplitter{
    index = this._getFirstIndex()

    _getFirstIndex = () => {
        let index  = this.word.find("//")
        if(index != -1)
            return index + 2
        else return 0
    }
    
    reset = (word, delimiter) => {
        this.word = `${word}${word[-1]==delimiter?"":delimiter}`
        this.index = this._getFirstIndex()
    }
}

const ArticleItems = {
    IMG: 0,
    HEADER: 1,
    TITLE: 2,
    DESCRIPTION: 3,
    AUTHOR: 4,
    LINK: 5,
    NEWS_INFO: 6,
    CODE: 7
}

function createArticleElementsTable(img=null, header=null, title=null, description=null, author=null){
    let obj = {}

    if(img) obj[ArticleItems.IMG] = img
    if(header) obj[ArticleItems.HEADER] = header
    if(title) obj[ArticleItems.TITLE] = title
    if(description) obj[ArticleItems.DESCRIPTION] = description
    if(author) obj[ArticleItems.AUTHOR] = author

    return obj
}

class LabeledObject{
    constructor(label, obj){
        this.label = label
        this.object = obj
    }
}

class DatabaseInfo{
    constructor(path, databaseFile, browser){
        this.path = path
        this.databaseFile = databaseFile
        this.browser = browser
    }
    
    get_path = () => window.path_join(this.path, this.databaseFile);
}

const ScheduleOperators = {
    null: 0,
    TITLE: 1,
    TIME: 2,
    DATE: 3,
    INTERVAL: 4,
    ACTION: 5
}

const ScheduleActions = {
    ADD: 0,
    DEL: 1,
    MOD: 2,
    null: 3
}
                    
function createScheduleObj(title=null, date=null, startTime=null, endTime=null){
    return {
        [ScheduleOperators.TITLE]: title,
        [ScheduleOperators.DATE]: date,
        [ScheduleOperators.TIME]: [startTime, endTime]
    }
}

class LinkedListNode{
    constructor(value, nextNode=null){
        this.value = value
        this.nextNode = nextNode
    }
}

class LinkedList{
    head = null
    tail = null
    length = 0
    currentNode = null

    constructor(firstElement = null){
        if(firstElement) this.push(firstElement)
    }

    push = (element) => {
        let node = new LinkedListNode(element)

        if(!this.head) this.head = this.tail = node
        else{
            this.tail.nextNode = node
            this.tail = node
        }

        this.length++
    }

    pop = () => {
        let node = this.head

        if(this.length<2){
            if(this.length === 1){
                this.head = this.tail = null
            }return node
        }else{
            node = this.head
            let lastNode

            while(node.nextNode){
                lastNode = node
                node = node.nextNode
            }

            this.tail = lastNode
            lastNode.nextNode = null
        }

        this.length--
        return node
    }

    shift = () => {
        if(this.head){
            let node = this.head
            this.head = this.head.nextNode
            this.length--

            return node
        }else return null
    }

    next = () => {
        if(this.currentNode) this.currentNode = this.currentNode.nextNode
        else this.currentNode = this.head

        return this.currentNode? this.currentNode.value: null
    }

    toArray = () => {
        let arr = []
        let node = this.head

        while(node){
            arr.push(node.value)
            node = node.nextNode
        }

        return arr
    }

    get = (index) => {
        let node

        if(index<this.length && index>-1){
            node = this.head
            for(let i=0; i<index; i++) node = node.nextNode
            node = node.value
        }

        return node
    }
}

class TaskManager{
    tasks = {}

    constructor(id){
        this.id = id
    }

    addTask = (func) => {
        
    }
}

export { Counter, FrequencyCounter, ArticleTagCounter, InterpreterEnvironment, WebsiteInfo, NewsWebsiteInfo, 
    NewsWebsitesInfo, LazySplitter, URLLazySplitter, ArticleItems, createArticleElementsTable, LabeledObject, DatabaseInfo,
    ScheduleActions, ScheduleOperators, createScheduleObj, LinkedList }